Api for mapping widget

Edru2 opened the pull request:
<!-- To keep things simple, focus on just one topic in your PR and make changes just for that topic.
     This'll make it a lot easier to review, and thus your PR will get in faster.
     Remember can open multiple PRs at a time! -->
#### Brief overview of PR changes/additions
Adds commands to open a MapWidget.
**openMapWidget()** -> opens MapWidget similar to clicking the Icon
**openMapWidget(dockingArea)** -> opens MapWidget and puts the widget in the chosen dockingArea. dockingArea is a string there are "f" floating "t" top "b" bottom "r" right and "l" left to choose
**openMapWidget(x,y,width,height)->** creates floating map Widget with given coordinates/sizes as numbers
**resizeMapWidget(width, height)** -> resizes MapWidget with given constraints
**moveMapWidget(x, y)** -> moves MapWidget to position x,y
**closeMapWidget()** -> closes (hides) the MapWidget

Geyser Wrapper for mentioned functions are added to Geyser.Mapper example:
**myMapWidget = Geyser.Mapper:new({embedded= false})**
Geyser constraints are:
**bool embedded** --is this an embedded mapper or a map widget?
**string dockPosition** --docking position at creation

+ if embedded is set to false at creation the mapper will be created as map window.
+ If embedded is set to true or is not set at all the mapper will be created as embedded mapper. 

+ If only embedded is set ( and no x, y, width, height values ) the mapping window will use the saved layout at creation if there is one.

+ If dockPosition is set to something other then floating the x, y, width, height values will be ignored (as we cannot set them in docking mode) and the MapWindow will be set into the dockPosition.

Geyser functions for the map window are
**myMapWidget:setDockPosition(dockPosition)**
#### Motivation for adding to Mudlet
https://github.com/Mudlet/Mudlet/issues/1256
#### Other info (issues closed, discussion etc)
If this works well I'm thinking about adding this functions (choose docking site...) also to UserWindows

Edru2 committed:
lua openMapWidget()

Adds commands to open a MapWidget.
openMapWidget() -> opens MapWidget similar to clicking the Icons
openMapWidget(dockingArea) -> opens MapWidget and puts the widget in the chosen dockingArea. dockingArea is a string there are "f" floating "t" top "b" bottom "r" right and "l" left to choose
openMapWidget(x,y,width,height)-> creates floating map Widget with given coordinates/sizes as numbers
resizeMapWidget(width, height) -> resizes MapWidget with given constraints
moveMapWidget(x,y) -> moves MapWidget to position x,y
closeMapWidget() -> closes (hides) the MapWidget

Geyser Wrapper for mentioned functions are added to Geyser.Mapper example:
myMapWidget = Geyser.Mapper:new({mapWidget = true})
Geyser constraints are:
bool mapWidget --is this mapper a map widget?
bool docked --start docked
string dockPos --docking position at creation

if one of those are set at creation the mapper will be craeted as map widget.
Geyser functions for the mapwidget are
myMapWidget:setDockPos(dockPos)

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2408,6 +2408,81 @@ bool mudlet::moveWindow(Host* pHost, const QString& name, int x1, int y1)
+        }
+        return {true, QString()};
+    } else {
+        if (area == "r") {
```
I'd use full names here - better readability and easier for beginners.

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserMapper.lua:
```diff
@@ -18,19 +18,57 @@ Geyser.Mapper = Geyser.Window:new({
+  end
+end
+
+function Geyser.Mapper:setDockPos(pos)
```
```suggestion
function Geyser.Mapper:setDockPosition(pos)
```

vadi2 reviewed (commented):
Looks really good at an overview - will test soon.

Thanks a lot for adding this!

Edru2 committed:
Added Suggestions and made dockposition case insensitive

Edru2 committed:
CodeFactor wanted a space

vadi2 commented on src/mudlet-lua/lua/geyser/GeyserMapper.lua:
```diff
@@ -49,8 +87,26 @@ function Geyser.Mapper:new (cons, container)
+    createMapper(me.windowname, me:get_x(), me:get_y(),
+    me:get_width(), me:get_height())
+  else
+    me.mapWidget = true
```
Seems to be a little bit of a bug around the mapWidget - it's possible to spawn two map widgets, one embedded and one floating, by doing:

```lua
mapWidget = Geyser.Mapper:new({mapWidget = false})
```

And then changing it to `true` and re-running the script.

Could I recommend a name better than `mapWidget`? It's not intuitive to remember whenever false or true means floating or embedded state. How about using a string "floating" or "embedded", and defaulting to "embedded" as the previous API did?

It's also a bit confusing to, for example, create a floating map widget - trying combinations of mapWidget and docked didn't yield a result. Mostly comes down to 'mapWidget' being a bit of an opaque name and not meaning much.

Let me know what you think!

Edru2 replied:
> Seems to be a little bit of a bug around the mapWidget - it's possible to spawn two map widgets, one embedded and one floating, by doing:

> mapWidget = Geyser.Mapper:new({mapWidget = false})

And then changing it to true and re-running the script.

Thats not really a new bug (or feature ;) depends how you see it). You can do the same already in 4.6.1 by first open a map window by clicking on the map icon and then use createMapper() for creating a embedded mapper.

> Could I recommend a name better than mapWidget? It's not intuitive to remember whenever false or true means floating or embedded state. How about using a string "floating" or "embedded", and defaulting to "embedded" as the previous API did?

I agree with you ( I'm not good at variable names) 
How about using a boolean embedded which is normally true and if I set that one to false at creation it creates the map window and if it's true (or not set) the embedded mapper.
+ I think I will get rid of docked as it is just doing the exactly same as dockPosition ="right".
I just added it because I was using it also for the UserWindow.

vadi2 replied:
Yeah, let's try those suggestions out and see how they feel.

vadi2 committed:
Added documentation links

vadi2 committed:
Optimised string comparison

vadi2 reviewed (commented):
Looks good, just a few points of feedback.

Edru2 committed:
Change variable name and get rid of docked

Change mapWidget to embedded.
docked is not really needed so I got rid of it.
embedded not has to be false to create a mapwindow to be less confusing.

vadi2 reviewed (approved):
Nice, that feels better and it's good to me.

Anyone from @Mudlet/lua-interface want to review?

demonnic commented:
Taking a look at this today, will :+1:  or :-1: in a bit

demonnic commented:
Looks good to me. My only comment is should we consider setting embedded to false if dockingPosition is provided, since it wouldn't be used for an embedded mapper?  I'm on the fence and could go either way.

![image](https://user-images.githubusercontent.com/3660/76685370-756bb600-65e9-11ea-88a6-b339a0b9d98d.png)

Edru2 commented:
> My only comment is should we consider setting embedded to false if dockingPosition is provided, since it wouldn't be used for an embedded mapper? I'm on the fence and could go either way.
> 
I had it that way before and then I started thinking it makes more sense this way. I'm not sure myself.

vadi2 commented:
Agree with @demonnic - since it's not embedded, the status should reflect that.

demonnic commented:
I mean, I'm not against needing to have both set intentionally for it to work, just thought it might be easier for those who want a docking position other than right or where it was left last not to have to provide both.

vadi2 commented:
I thought it was implicit already if you set a docking position so I was talking about something else.

Edru2 committed:
make dockPosition create an map window again

Edru2 commented:
I made it the previous way again and it make more sense to me now.
If someone only set dockPosition it will be map window again,
but if someone creates a mapper and sets embedded=true and a dockPosition, the mapper will be embedded ;) and dockPosition ignored.

vadi2 commented:
We shook it down to something that works well :)

vadi2 commented:
@Edru2 One last thing to go - can you document all of this? Say that it'll be available in the upcoming Mudlet 4.7 update for the usual note field.

SlySven commented:
Hang on - I thought I had mentioned this before (perhaps it was for a different PR) but `QDockWidget`s, including a mapper that may be docked, need to have a dock area specified even if they are created in a floating state - so that when the dock button is pressed they know which dock area to dock into. Basically you need BOTH a dock area argument AND a docked/floating argument.

As I think I said in a previous comment we only get away with it for toolbars because specifying a left, top or right dock area produces a `TEasyButtonBar` that is always fixed to that part of the main `TConsole` (not the Application's central widget) and specifying a `floating` toolbar produces a `TToolBar` that is indeed floating (but is set to dock to the `Qt::LeftDockWidgetArea` when created).

vadi2 commented:
Are you certain? Try the PR out - this is handled on the Lua level.

Edru2 committed:
added QStringliteral

thanks for pointing this out @SlySven
I hope I didn't forget this somewhere else.

Edru2 commented:
> Hang on - I thought I had mentioned this before (perhaps it was for a different PR) but `QDockWidget`s, including a mapper that may be docked, need to have a dock area specified even if they are created in a floating state - so that when the dock button is pressed they know which dock 

This PR doesn't change anything from the map window creation code, so even if it isn't mentioned anywhere there is always a dock area specified at creation (in this case Qt::RightDockWidgetArea).
